### PR TITLE
Fix Windows path handling and all tests on Windows

### PR DESCRIPTION
**Pull request by aRestless**
This PR primarily addresses #572, but also fixes all other tests on Windows. Summary of the important changes:

- In logstream.go and tail.go, when the URL protocol cannot be recognized, the whole argument is now treated as a path. That leads to supporting file paths like `C:\somefile`. It should be noted though that local file paths and URLs are two separate concepts that are incompatible in edge cases, even on Linux, since colons are valid characters for file names. So it might be worth it to separate these two use cases in the future.
- File handles in tests are now explicitly closed, otherwise the temp folder cleanup in fs.go will not be able to delete the files on Windows
- All tests regarding pipes and sockets have been moved to Unix-only tests. It might be possible to get at least some of them running on Windows in the future.
- All tests relying on "permission denied" errors have been moved to Unix-only tests. On Windows, it is not possible to create a file that you cannot read yourself (minimum file permissions are 0222).

### CODE REVIEW

File: internal/mtail/multiple_levels_directory_integration_test.go [hunk 1]
```diff
@@ -4,13 +4,11 @@
 package mtail_test
 
 import (
+	"github.com/google/mtail/internal/mtail"
+	"github.com/google/mtail/internal/testutil"
```
Comment (jaqx0r) [inline internal/mtail/multiple_levels_directory_integration_test.go:8]: Put the local imports last please.

File: internal/mtail/multiple_levels_directory_integration_test.go [hunk 2]
```diff
@@ -28,9 +26,10 @@ func TestPollLogPathPatterns(t *testing.T) {
 	lineCountCheck := m.ExpectExpvarDeltaWithDeadline("lines_total", 1)
 
 	logFile := filepath.Join(logDir, "files", "a", "log", "a.log")
-	testutil.FatalIfErr(t, os.MkdirAll(path.Dir(logFile), 0700))
+	testutil.FatalIfErr(t, os.MkdirAll(filepath.Dir(logFile), 0700))
 
 	f := testutil.TestOpenFile(t, logFile)
+	defer f.Close()
```
Comment (jaqx0r) [inline internal/mtail/multiple_levels_directory_integration_test.go:32]: I think TestOpenFile should register a t.Cleanup() function to handle the Close instead of deferring in test here, but I'll take care of that in another PR.

File: internal/mtail/permission_denied_integration_unix_test.go [hunk 3]
```diff
@@ -12,6 +15,8 @@ import (
 	"github.com/google/mtail/internal/testutil"
 )
 
+// TestPermissionDeniedOnLog is a unix-specific test because on Windows, it is not possible to create a file
+// that you yourself cannot read (minimum permissions are 0222).
```
Comment (jaqx0r) [inline internal/mtail/permission_denied_integration_unix_test.go:19]: Thanks for documenting that!

**Review by jaqx0r — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Thanks for documenting why the tests are unix specific.

I have a question about log rotation then -- on Windows is the standard practice to truncate the file?  Or do logs never get rotated?  A file can't be renamed if it's open?

I didn't completely check but it looks like the LogTruncation test just happens to test the Windows behaviour.  Is there any other way that Windows log files are handled that we should cover?  I just feel that the integration tests have reduced coverage.

I'm not sure if you also want to make sure the end to end tests run on Windows, just to make sure that we cover any log-tailing bugs -- this test has helped me find errors connecting the tailer to the VM in the past.  

The only other thing is that the PR needs rebasing or merging with master, and the lint errors need addressing.

Thanks again!

File: internal/mtail/multiple_levels_directory_integration_test.go (hunk 2 above)
Comment (aRestless) [reply internal/mtail/multiple_levels_directory_integration_test.go:32]: Very good idea!

**Comment by aRestless** [comment]
> I have a question about log rotation then -- on Windows is the standard practice to truncate the file? Or do logs never get rotated? A file can't be renamed if it's open?

I also don't see myself as an expert but the practice I've seen is that indeed, the current log file is being copied and then truncated. In a similar manner, on Windows you also can't truncate a file that you've opened in append mode. That is why I switched from `TestOpenFile` to `OpenLogFile` in `TestHandleLogTruncate`.

> I didn't completely check but it looks like the LogTruncation test just happens to test the Windows behaviour. Is there any other way that Windows log files are handled that we should cover? I just feel that the integration tests have reduced coverage.

I would agree that the LogTruncation tests cover the Windows use case. The only other oddity I know of are the FILE_SHARE_READ and FILE_SHARE_WRITE "share mode" flags when opening files on Windows. They allow other processes to still access the file. Testing access to files opened without those flags could in theory be a substitute for the tests relying on permission denied errors for Linux. However, Golang _always_ sets those flags, so it's not possible to simulate from within a Golang test.

> I'm not sure if you also want to make sure the end to end tests run on Windows, just to make sure that we cover any log-tailing bugs -- this test has helped me find errors connecting the tailer to the VM in the past.

I feel immensely blind but can you point me towards where the end to end tests are defined? But regardless, that might be something for a future PR.

> The only other thing is that the PR needs rebasing or merging with master, and the lint errors need addressing.

Will get on that now. Thanks!

**Comment by aRestless** [comment]
Linting errors should be fixed. I shot gofumpt/gofmt against everything and that has turned all octal notations to the form 0o666. Please let me know if that's alright or if that should be reverted.

Once all checks are green I'd rebase and squash for an easy merge and a clean history.

File: internal/mtail/multiple_levels_directory_integration_test.go (hunk 1 above)
Comment (aRestless) [reply internal/mtail/multiple_levels_directory_integration_test.go]: done

**Comment by jaqx0r** [comment]
> > I have a question about log rotation then -- on Windows is the standard practice to truncate the file? Or do logs never get rotated? A file can't be renamed if it's open?
> 
> I also don't see myself as an expert but the practice I've seen is that indeed, the current log file is being copied and then truncated. In a similar manner, on Windows you also can't truncate a file that you've opened in append mode. That is why I switched from `TestOpenFile` to `OpenLogFile` in `TestHandleLogTruncate`.

Ok, so in Windows, a process's log library would handle copying its own log file, and then doing the truncation.  Or is it an external process that would do this?

When you say you've seen this practice, can you give me an example of an application that does this?

> > I didn't completely check but it looks like the LogTruncation test just happens to test the Windows behaviour. Is there any other way that Windows log files are handled that we should cover? I just feel that the integration tests have reduced coverage.
> 
> I would agree that the LogTruncation tests cover the Windows use case. The only other oddity I know of are the FILE_SHARE_READ and FILE_SHARE_WRITE "share mode" flags when opening files on Windows. They allow other processes to still access the file. Testing access to files opened without those flags could in theory be a substitute for the tests relying on permission denied errors for Linux. However, Golang _always_ sets those flags, so it's not possible to simulate from within a Golang test.

Intriguing.  Thanks for the details.  We can come back to this comment thread in the future for reference.  (I'm wondering about skipping the file abstraction library and going straight into Windows syscalls if we need to.)

I'm satisfied with not having the permission denied test in Windows as it was an edge case that showed up during deployment of a very early version of `mtail`.

I think my feeling of coverage is entirely because I don't understand file semantics on windows and I don't know how log rotation works there either :)


 
> > I'm not sure if you also want to make sure the end to end tests run on Windows, just to make sure that we cover any log-tailing bugs -- this test has helped me find errors connecting the tailer to the VM in the past.
> 
> I feel immensely blind but can you point me towards where the end to end tests are defined? But regardless, that might be something for a future PR.

One of the tests you edited is called examples_integration_test, which you've made a `+build linux` only test.  If you can replicate the file-read part, the rest should be identical between linux and windows versions.  Happy to see it in a followup.

File: internal/mtail/basic_tail_integration_test.go [hunk 4]
```diff
@@ -4,52 +4,14 @@
...
 	"os"
 	"path/filepath"
-	"sync"
 	"testing"
 
 	"github.com/google/mtail/internal/mtail"
 	"github.com/google/mtail/internal/testutil"
 )
 
-func TestBasicTail(t *testing.T) {
```
Comment (jaqx0r) [inline internal/mtail/basic_tail_integration_test.go:17]: If this test can be made to work on Windows, that would be ideal.  It really just checks that the tailer can open a file, which I expect should also work on Windows.

File: internal/mtail/examples_integration_unix_test.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
+// Copyright 2011 Google Inc. All Rights Reserved.
+// This file is available under the Apache license.
+
+//go:build unix
+// +build unix
+
+package mtail_test
+
+// Two mtails both alike in dignity.
```
Comment (jaqx0r) [inline internal/mtail/examples_integration_unix_test.go:9]: Pop a note here like you did for the other tests saying that Pipes are unix only thing.

File: internal/mtail/golden/reader_test.go [hunk 6]
```diff
@@ -106,7 +106,7 @@ var expectedMetrics = metrics.MetricSlice{
 		LabelValues: []*metrics.LabelValue{
 			{
 				Labels: []string{},
-				Value:  datum.MakeString("hi", time.Date(2018, 6, 16, 18, 04, 0, 0, time.UTC)),
+				Value:  datum.MakeString("hi", time.Date(2018, 6, 16, 18, 0o4, 0, 0, time.UTC)),
```
Comment (jaqx0r) [inline internal/mtail/golden/reader_test.go:109]: This one's a bug!  Just make it a decimal 4 please.

File: internal/runtime/vm/vm_test.go [hunk 7]
```diff
@@ -789,7 +789,7 @@ func TestStrptimeWithTimezone(t *testing.T) {
 	vm.t.Push("2012/01/18 06:25:00")
 	vm.t.Push("2006/01/02 15:04:05")
 	vm.execute(vm.t, obj.Program[0])
-	if vm.t.time != time.Date(2012, 01, 18, 06, 25, 00, 00, loc) {
+	if vm.t.time != time.Date(2012, 0o1, 18, 0o6, 25, 0o0, 0o0, loc) {
```
Comment (jaqx0r) [inline internal/runtime/vm/vm_test.go:792]: Also bugs, just plain decimals here please.

File: internal/runtime/vm/vm_test.go [hunk 8]
```diff
@@ -802,7 +802,7 @@ func TestStrptimeWithoutTimezone(t *testing.T) {
 	vm.t.Push("2012/01/18 06:25:00")
 	vm.t.Push("2006/01/02 15:04:05")
 	vm.execute(vm.t, obj.Program[0])
-	if vm.t.time != time.Date(2012, 01, 18, 06, 25, 00, 00, time.UTC) {
+	if vm.t.time != time.Date(2012, 0o1, 18, 0o6, 25, 0o0, 0o0, time.UTC) {
```
Comment (jaqx0r) [inline internal/runtime/vm/vm_test.go:805]: and here.

**Review by jaqx0r — COMMENTED** [review]
I'm sorry about the octal changes.  They're fine, except that it's highlighted some places where they should not have been interpreted as octal.

**Comment by aRestless** [comment]
> Ok, so in Windows, a process's log library would handle copying its own log file, and then doing the truncation. Or is it an external process that would do this?
> 
> When you say you've seen this practice, can you give me an example of an application that does this?

An equivalent to "logrotate" as a standalone tool is very rare on Windows. One example of a reimplementation of the logrotate API for Windows can be found [here](https://github.com/plecos/logrotatewin). Logrotate does support the "copytruncate" mode, which is the one I was describing. But again, this is rarely used, and logs on Windows are often an afterthought.

My use case at the moment is Unreal Engine (and with that, all games built on it), which does the copy-then-truncate method on startup.

> One of the tests you edited is called examples_integration_test, which you've made a +build linux only test. If you can replicate the file-read part, the rest should be identical between linux and windows versions. Happy to see it in a followup.

That's a mistake then. All platform-specific tests should have received a proper file name suffix (sadly it also needs the build tag in the file, because _test and _platform suffixes don't work both on one file). Will rectify that and check if the test can be run on Windows.

File: internal/mtail/basic_tail_integration_test.go (hunk 4 above)
Comment (aRestless) [reply internal/mtail/basic_tail_integration_test.go:17]: I think this ended up in the other file on accident. Will be reverted on the next push.

File: internal/mtail/examples_integration_unix_test.go (hunk 5 above)
Comment (aRestless) [reply internal/mtail/examples_integration_unix_test.go:10]: done

File: internal/runtime/vm/vm_test.go (hunk 7 above)
Comment (aRestless) [reply internal/runtime/vm/vm_test.go]: done

File: internal/runtime/vm/vm_test.go (hunk 8 above)
Comment (aRestless) [reply internal/runtime/vm/vm_test.go]: done

File: internal/mtail/golden/reader_test.go (hunk 6 above)
Comment (aRestless) [reply internal/mtail/golden/reader_test.go]: done

**Comment by aRestless** [comment]
I think the only concern left is the end to end tests. `TestExamplePrograms` is included, but that tests oneshot, not actually streamed data. Would you say there is value in an additional test that copies the content of the source file into the log file that is being read?

**Comment by jaqx0r** [comment]
> I think the only concern left is the end to end tests. `TestExamplePrograms` is included, but that tests oneshot, not actually streamed data. Would you say there is value in an additional test that copies the content of the source file into the log file that is being read?

Ultimately yes I do.  But, that's perhaps a PR for another day?

**Review by jaqx0r — APPROVED** [review]

**Comment by jaqx0r** [comment]
Looks like the linter is unhappy again, please fmt again with your rebase, thanks!

**Comment by aRestless** [comment]
Fixed the linting errors, rebased and squashed. I hope everything is alright now, I had some issues with git on Windows (file endings and all that).

For the additional test I've got something around 90% finished and would likely need your input, but as you said that's something for a different PR.

**Review by jaqx0r — APPROVED** [review]

**Comment by jaqx0r** [comment]
Thanks for working on this!
